src/include/function/gds/gds_object_manager.h

```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include "storage/buffer_manager/memory_manager.h"
+
+namespace kuzu {
+namespace function {
+
+// ObjectBlock represents a pre-allocated amount of memory that can hold up to maxElements objects
+// ObjectBlock should be accessed by a single thread.
+template<typename T>
+class ObjectBlock {
+public:
+    ObjectBlock(std::unique_ptr<storage::MemoryBuffer> block, uint64_t sizeInBytes)
+        : block{std::move(block)} {
+        maxElements.store(sizeInBytes / (sizeof(T)), std::memory_order_relaxed);
+        nextPosToWrite.store(0, std::memory_order_relaxed);
+    }
+
+    T* reserveNext() {
+        auto result = getData() + nextPosToWrite.load(std::memory_order_relaxed);
+        nextPosToWrite.fetch_add(1, std::memory_order_relaxed);
+        return result;
+    }
+
+    bool hasSpace() const {
+        return nextPosToWrite.load(std::memory_order_relaxed) <
+               maxElements.load(std::memory_order_relaxed);
+    }
+
+private:
+    T* getData() const { return reinterpret_cast<T*>(block->buffer.data()); }
+
+private:
+    std::unique_ptr<storage::MemoryBuffer> block;
+    std::atomic<uint64_t> maxElements;
+    std::atomic<uint64_t> nextPosToWrite;
+};
+
+// ObjectArrayByTable represents a pre-allocated amount of object per tableID.
```

Comment:
This sounds wrong because it makes it sound like this is an Array and not a Map. Also "ByTable" sounds awkward to me. I'd call it TableToObjectArrayMap or ObjectArraysMap and we don't have to clear what the key is.

Reply:
Done